File: 3rd_party/adios/python/adios2/adios.py
"""License:
  Distributed under the OSI-approved Apache License, Version 2.0.  See
  accompanying file Copyright.txt for details.
"""

from adios2 import bindings
from adios2.io import IO
from adios2.operator import Operator


class Adios:
    """High level representation of the ADIOS class in the adios2.bindings"""

    def __init__(self, config_file=None, comm=None):
        if comm and not bindings.is_built_with_mpi:
            raise RuntimeError("Cannot use MPI since ADIOS2 was built without MPI support")

        if config_file:
            if comm:
                self.impl = bindings.ADIOS(config_file, comm)
            else:
                self.impl = bindings.ADIOS(config_file)
        else:
            if comm:
                self.impl = bindings.ADIOS(comm)
            else:
                self.impl = bindings.ADIOS()

    @property
    def impl(self):
        """Bindings implementation of the class"""
        return self._impl

    @impl.setter
    def impl(self, implementation):
        self._impl = implementation

    def declare_io(self, name):
        """
        Declare IO instance

        Args:
            name (str): IO instance name
        """
        return IO(self.impl.DeclareIO(name), name, self)

    def at_io(self, name):
        """
        Inquire IO instance

        Args:
            name (str): IO instance name
        Returns:
            IO: IO instance
        """
        io = None
        io_instance = self.impl.AtIO(name)
        if io_instance:
            io = IO(io_instance, name, self)
        return io

    def remove_io(self, name):
        """
        Remove IO instance

        Args:
            name (str): IO instance name
        """
        self.impl.RemoveIO(name)

    def remove_all_ios(self):
        """Remove all IO instance"""
        self.impl.RemoveAllIOs()

    def define_operator(self, name, kind, parameters={}):
        """
        Add an operation (operator).

        Args:
            name (str): name of the operator.
            kind (str): name type of the operation.
            params (dict): parameters as a form of a dict for the operation.
        """
        return Operator(self.impl.DefineOperator(name, kind, parameters), name)

    def inquire_operator(self, name):
        """
        Add an operation (operator).

        Args:
            name (str): name of the operator.
        Return:
            Operator: requested operator
        """
        operator = None
        operator_instance = self.impl.InquireOperator(name)
        if operator_instance:
            operator = Operator(self.impl.InquireOperator(name), name)

        return operator

    def flush_all(self):
        """
        Flush all IO instances
        """
        self.impl.FlushAll()

File: 3rd_party/adios/bindings/CXX11/adios2/cxx11/ADIOS.cpp
/* Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * ADIOS.cpp : public ADIOS class using PIMPL for C++11 bindings
 * Created on: Jun 4, 2018
 *     Author: William F Godoy
 */

#include "ADIOS.h"

#include "adios2/core/ADIOS.h"
#include "adios2/core/IO.h"
#include "adios2/core/VariableStruct.h"
#include "adios2/helper/adiosFunctions.h" //CheckForNullptr

namespace adios2
{
ADIOS::ADIOS(const std::string &configFile)
: m_ADIOS(std::make_shared<core::ADIOS>(configFile, "C++"))
{
}

ADIOS::ADIOS() : ADIOS("", "C++") {}

ADIOS::ADIOS(const std::string &configFile, const std::string &hostLanguage)
: m_ADIOS(std::make_shared<core::ADIOS>(configFile, hostLanguage))
{
}

ADIOS::operator bool() const noexcept { return m_ADIOS ? true : false; }

IO ADIOS::DeclareIO(const std::string name, const ArrayOrdering ArrayOrder)
{
    CheckPointer("for io name " + name + ", in call to ADIOS::DeclareIO");
    return IO(&m_ADIOS->DeclareIO(name, ArrayOrder));
}

IO ADIOS::AtIO(const std::string name)
{
    CheckPointer("for io name " + name + ", in call to ADIOS::AtIO");
    return IO(&m_ADIOS->AtIO(name));
}

void ADIOS::FlushAll()
{
    CheckPointer("in call to ADIOS::FlushAll");
    m_ADIOS->FlushAll();
}

void ADIOS::EnterComputationBlock() noexcept
{
    CheckPointer("in call to ADIOS::EnterComputationBlock()");
    m_ADIOS->EnterComputationBlock();
}

void ADIOS::ExitComputationBlock() noexcept
{
    CheckPointer("in call to ADIOS::ExitComputationBlock()");
    m_ADIOS->ExitComputationBlock();
}

Operator ADIOS::DefineOperator(const std::string name, const std::string type,
                               const Params &parameters)
{
    CheckPointer("for operator name " + name + ", in call to ADIOS::DefineOperator");
    auto op = &m_ADIOS->DefineOperator(name, type, parameters);
    return Operator(op->first, &op->second);
}

Operator ADIOS::InquireOperator(const std::string name)
{
    CheckPointer("for operator name " + name + ", in call to InquireOperator");
    auto op = m_ADIOS->InquireOperator(name);
    if (op == nullptr)
    {
        return Operator("", nullptr);
    }
    else
    {
        return Operator(op->first, &op->second);
    }
}

bool ADIOS::RemoveIO(const std::string name)
{
    CheckPointer("for io name " + name + ", in call to ADIOS::RemoveIO");
    return m_ADIOS->RemoveIO(name);
}

void ADIOS::RemoveAllIOs() noexcept
{
    CheckPointer("in call to ADIOS::RemoveAllIOs");
    m_ADIOS->RemoveAllIOs();
}

// PRIVATE
void ADIOS::CheckPointer(const std::string hint)
{
    if (!m_ADIOS)
    {
        throw std::invalid_argument("ERROR: invalid ADIOS object, did you call "
                                    "any of the ADIOS explicit "
                                    "constructors?, " +
                                    hint + "\n");
    }
}

} // end namespace adios2
